Add support for placing JavaScript libraries in static libs

NWilson opened the pull request:
This patch makes our build process massively, massively simpler!

Our code is split into a few dozen libraries, and some of them now have accompanying files needing `--js-library`. This is a problem for our cmake setup, which can't easily pass the filenames through from a base library to the final link commandline.

This commit makes emscripten behave a bit more like a standard compiler, but allowing `.jso` files to be "compiled" to `.jso.o` files and included in static libraries. They get used at the link stage.

```
emcc myfile.c -c -o myfile.c.o
# pretend to "compile" it to make the build system happy, really just copies it:
emcc myjslib.jso -c -o myjslib.jso.o
emar rc libtest.a myfile.c.o myjslib.jso.o
emcc app.c -c -o app.c.o
# uses the js files inside libtest.a as if passed with --js-library
emcc -o app.html app.c.o libtest.a
```

This also makes dependency management for incremental builds very simple since it looks to the build system just like a standard source file. Much nicer than the hacky dependency stuff in the existing emscripten cmake module.

If you have an idea for implementing this sort of build setup without this extension I'd be grateful, but this seems like the path of least resistance to me!

NWilson commented on emcc:
```diff
@@ -1275,6 +1288,9 @@ try:
 
   # Emscripten
   logging.debug('LLVM => JS')
+  for lib in library_files:
```
OK, I've done some perf testing for you to check this on a couple of sub-projects of our various products which use emscripten.

I've used three build options:

1. Stock emscripten (incoming). All jso files excluded from the build/link because it doesn't understand them
1. Modified emscripten (this pull request). All jso files excluded from the build/link for comparison purposes
1. Modified emscripten, compiling and linking in the our jso files as well

All of these are using plain make, "-j 1", clean builds, timed after running CMake to set up the makefiles. The machine has an SSD.

### Project 1 - Debug build, 338 C++ files, 375 headers, uses a couple of static libraries

option | times | average
-------|-------|--------
stock emscripten, jso excluded | 1:09.0, 1:08.8, 1:09:2 | 69.0
new emscripten, jso excluded | 1:09.6, 1:09.2, 1:08.8 | 69.2
new emscripten, jso included | 1:10.7, 1:10.2, 1:10.0 | 70.3

### Project 2 - RelWithDebInfo build, ~600 C++ files, uses quite a few static libraries

option | times | average
-------|-------|--------
stock emscripten, jso excluded | 2:37.6, 2:37.0, 2:37.5 | 157.4
new emscripten, jso excluded | 2:37.7, 2:37.1, 2:37.4 | 157.4
new emscripten, jso included | 2:38.8, 2:38.8, 2:37.9 | 158.5

### Conclusion

Adding jso files significantly increases link time! That's actually a bit more of an effect than I'd anticipated. But, there's no significant increase in time with/without these changes for projects that aren't using JS libraries.

NWilson commented on tools/shared.py:
```diff
@@ -1637,35 +1604,94 @@ def closure_compiler(filename, pretty=True):
-      b = open(filename, 'r').read(24)
-      return b[20] == 'B' and b[21] == 'C'
+    try:
+      if filename in Building._is_bitcode_cache:
```
OK, I've done some perf testing on the three caches that are involved in these commits. Some slightly less scientific testing than the runs I did above, just to see if the caching has any effect at all. I didn't do as many repeats for each option here.

### High-level benchmarking
#### 300 sources compile

option | time
-------|-----
all caches present | 2:22:3
all caches removed | 2:21.7 (17.3 system)
is_ar cache only | 2.22:8 (17.4 system)
is_bitcode cache only | 2:23.5 (17.6 system)
ar_contents cache only | 2:21.4 (16.9 system)

#### Link a single output file

option | time
-------|-----
all caches present | 14.5 (1.4 system)
all caches removed | 14.4 (1.4 system)
is_ar cache only | 14.3 (1.4 system)
is_bitcode cache only | 14.3 (1.3 system)
ar_contents cache only | 14.4 (1.4 system)

#### Conclusion

Hard to be sure from the high-level overall timings, the margins are so small! The tables above are basically useless, the effect of the caches is completely swamped by noise.

### Some more micro-benchmarking of the specific functions

It appears that the `ar_contents` cache is useful, giving a consistent 200ms reduction in disc access time (system time). It saves calling LLVM `ar` more than half the time during the link stage. I'd keep this one in.

The `is_ar` cache is useful too I think; over half the time it kicks in and prevents a disk access (specifically, around 40 cache hits, 30 cache misses for typical link lines throughout our projects). I can't measure any timing impact of that at all though!

The `is_bitcode` cache is completely useless, each bitcode file is only queried once for all the compilation and link commands in our project.

NWilson commented on tools/shared.py:
```diff
@@ -1637,35 +1604,94 @@ def closure_compiler(filename, pretty=True):
 
+  @staticmethod
+  def is_jslibrary(filename):
+    if not filename.endswith(('.jso','.jso.o')):
```
We support both `.jso` and `.jso.o` because CMake has an annoying habit of tacking ".o" onto the end of everything. If you tell it to treat `.jso` files as input files, it works as hard as it can to produce `.jso.o` output, and I couldn't find a way to stop it (it seems fairly baked in behaviour).

Rather than fight the system, I just made emcc shrug and handle `.jso.o`, which makes integrating JS libraries very easy into projects using CMake.

NWilson commented:
@kripken I'm back in action on this PR!

1. Rebased as requested
2. Running...
3. Test added now in 1b8b9fed2d8b912a5c53da03c517699f8bcda594

Outstanding things from the line-by-line comments:
* Performance concerns -- I've done a fairly thorough test, it doesn't add any time at all to compilation 
* Testing the various caches -- that was a very good spot you made! You're right that the caches aren't all that useful, I took out the useless one in bc31ad0.
* Concerns about the `.jso` file extension -- I've done some decent googling, I just can't find any software-related hits for that string at all. I think we're safe to use this extension.

NWilson commented:
@kripken ping?

In `runner.py other` and `runner.py asm1` there are some failures for me on incoming, but excluding them I've found two things upset by my changes here, `other.test_dup_o_in_one_a` and `other.test_symlink` (all the asm1 tests pass).

Assuming I can fix those tests, is this moving towards acceptance?

kripken commented on emcc:
```diff
@@ -808,7 +813,7 @@ try:
   has_dash_c = '-c' in newargs
   if has_dash_c:
-    assert has_source_inputs or has_header_inputs, 'Must have source code or header inputs to use -c'
+    assert has_source_inputs or has_header_inputs or has_jsobj_inputs, 'Must have source code or header inputs to use -c'
```
I am still puzzled on this matter.

NWilson replied:
Good question, OK, I'll try and explain myself better.

What we'd like is the following sequence of commands:

    emcc -o lib1.c.o -c lib1.c              # Uses function in libjs.jso
    emar cr libexample.a lib1.c.o libjs.jso # Make lib containing jso file
    emcc -o main.c.o -c main.c              # Uses function in lib1.c
    emcc -o main.js main.c.o libexample.a

And indeed that works! Unfortunately, it's a real pain getting a build system like CMake to run those commands, because CMake is designed to compile things, and there's no "compile" step for the file `libjs.jso`. It doesn't expect to use raw files in static libraries.

So, if you were compiling with the following CMakeLists.txt file you'd generate a slightly different set of build commands:

    cmake_minimum_required(VERSION 2.8)
    project(testproj)
    add_library(example STATIC lib1.c libjs.jso)
    add_executable(main main.c)
    target_link_libraries(main example)

CMake will run the following commands:

    emcc -o lib1.c.o -c lib1.c            #     Same as before
    emcc -o libjs.jso.o -c libjs.jso      # *** "compile" the jso file (raw copy)
    emar cr libexample.a lib1.c.o libjs.jso.o # Include libjs.jso.o
    emcc -o main.c.o -c main.c            #     Same as before
    emcc -o main.js main.c.o libexample.a #     Same as before

So the hack in `***` is fairly simple: we allow build systems to run `emcc -o libjs.jso.o -c libjs.jso` (equivalent to `cp libjs.jso libjs.jso.o`), treating jso files just like all the other source files. Incremental builds pick up the dependencies beautifully.

kripken replied:
Now I see, thanks

kripken commented on emcc:
```diff
@@ -1259,6 +1272,9 @@ try:
 
   # Emscripten
```
please move this comment down, and add a different comment here about jsos

NWilson replied:
Good point, done.

kripken commented on tools/shared.py:
```diff
@@ -1258,46 +1255,6 @@ def consider_object(f, force_add=False):
-        # will trample each other when we extract them. to help warn of such situations, we warn if there
-        # are duplicate entries in the archive
-        if len(contents) != len(set(contents)):
-          logging.warning('loading from archive %s, which has duplicate entries (files with identical base names). this is dangerous as only the last will be taken into account, and you may see surprising undefined symbols later. you should rename source files to avoid this problem (or avoid .a archives, and just link bitcode together to form libraries for later linking)' % f)
```
this is important code! not sure why it is getting killed?

NWilson replied:
It's not gone away, it's just been split out into a separate function further down the file (`ar_contents`). I didn't even make any changes to this, I just needed it to be an independent function so I could call it elsewhere.

(Actually, I did change it by mistake when rebasing, I lost a few lines of changes that had been made up here. I've put those back now.)

kripken commented:
Yes, I would say this is moving towards acceptance. I like the idea, and perf seems ok based on your tests. Still a bunch of issues left, though, I added some comments now.

We will also need to document this, not sure offhand where. Perhaps on the "building projects" page. Or maybe we need a new page for this?

NWilson commented:
Thanks for your help!

I've fixed up a few nits, and added some documentation.  There's now an example of how to build the jso files on "Building Projects / Using libraries", and more details on how to make your own JavaScript libraries in "Interacting with code / Implement a C API in JavaScript".

kripken commented on emcc:
```diff
@@ -1059,7 +1065,13 @@ try:
   for i, input_file in input_files:
     file_ending = filename_type_ending(input_file)
-    if file_ending.endswith(SOURCE_ENDINGS):
+    if file_ending.endswith(JSOBJ_ENDINGS):
```
How about putting this at the end of the if-elif chain, as it is the least important/clear to the typical reader.

kripken commented on emcc:
```diff
@@ -1257,9 +1270,15 @@ try:
 
   log_time('post-link')
 
+  # Prepare JS libraries for Emscripten linking, extracting and merging .jso
```
This part feels wrong to me. We have already processed these `.a` files and extracted their contents, haven't we? We should be able to just move the extracted files from object files to js library files, if they are `.jso`s.

kripken commented:
I am not sure how `.a` dependency tracking works here? Normally we use `llvm-nm` to scan object files in `.a`s, and pull in the ones that provide symbols that are yet undefined. What do we do for a `.jso`? Is it unconditionally pulled in?

Wouldn't `nm` fail on a `.jso` file?

kripken commented on tools/shared.py:
```diff
@@ -1627,20 +1584,85 @@ def is_ar(filename):
+    if not filename.endswith(('.jso','.jso.o')):
+      return False
+    try:
+      with open(filename, 'r') as f:
```
this part is very worrying to me. why do we need this hack?

i would think we should remove this and just check the suffixes in `emcc`, the only place this is used. was there a reason not to do that?

NWilson commented:
Closing... This was so close to being merged, after rounds of review, benchmarking, tests and website additions... but we're moving now to a standalone Clang+LLD toolchain for WebAssembly, which no longer needs Emscripten.
